### PR TITLE
Use fallback debug impls instead of unimplemented

### PR DESCRIPTION
**Pull request by jackh726**
Also turn on deny(rust_2018_idioms)

### CODE REVIEW

File: chalk-ir/src/debug.rs [hunk 1]
```diff
@@ -23,78 +23,63 @@ impl<I: Interner> Debug for AssocTypeId<I> {
 
 impl<I: Interner> Debug for Ty<I> {
     fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), Error> {
-        I::debug_ty(self, fmt)
-            .unwrap_or_else(|| unimplemented!("cannot format Ty without setting Program in tls"))
+        I::debug_ty(self, fmt).unwrap_or_else(|| write!(fmt, "Ty(?)"))
```
Comment (nikomatsakis) [inline chalk-ir/src/debug.rs]:
```suggestion
        I::debug_ty(self, fmt).unwrap_or_else(|| write!(fmt, "{:?}", self.interned))
```

Weren't we going to print the 'interned' format instead? Something like this?

File: chalk-ir/src/lib.rs [hunk 2]
```diff
@@ -1325,6 +1343,10 @@ impl<I: Interner> Goal<I> {
         Self { interned }
     }
 
+    pub fn interned(&self) -> &I::InternedGoal {
```
Comment (nikomatsakis) [inline chalk-ir/src/lib.rs]: I don't object to these methods, but we're not using them anywhere, right? And I don't really think they're *needed*

File: chalk-ir/src/debug.rs (hunk 1 above)
Comment (jackh726) [reply chalk-ir/src/debug.rs]: 🤦‍♀️I forgot those are `: Debug`

File: chalk-ir/src/lib.rs (hunk 2 above)
Comment (jackh726) [reply chalk-ir/src/lib.rs]: Well, I was using them in rustc for debug_* functions. Alternatively, we could pass `I::Interned*` to debug_* functions

**Pull request merged**